juju/juju
# state: prepare for persistent storage

## Description of change

We make various changes to the state package
to allow for volumes and filesystems to be
bound to a model, rather than only machine
or storage instance. We continue to bind
the storage to storage instances; we will
later flip the switch when commands for
detaching, attaching, and removing storage
have been introduced.

This is the minimal change. Volume-backed
filesystems are currently always considered
machine-scoped, which leads to undesirable
behaviour: the filesystem cannot be detached
from the machine it is initially attached to
(or doing so will cause the volume to be
destroyed). We will need to make further
changes so that volume-backed filesystems
are not scoped or bound to a single machine.

This is based on https://github.com/juju/juju/pull/6795,
with a couple of small fixes to the assertions; and
until we have commands for destroying storage, we
continue to bind machine storage to the storage
instance.

## QA steps

1. juju bootstrap localhost
2. juju deploy a charm with rootfs storage
3. juju remove-unit
4. ensure the machine storage is also removed
